### ISSUE TITLE
onlogout not being called in IE9 after navigator.id.watch with undefined loggedInUser

### ISSUE THREAD

**Issue by TeeTeeHaa** [mozilla/persona#3417]
Whether or not the "onlogout" callback is being called does differ in Internet Explorer 9 compared to Firefox 20 and Chrome 26. According to the documentation (https://developer.mozilla.org/en-US/docs/DOM/navigator.id.watch) the behavior in Internet Explorer 9 is wrong.

```
$(document).ready(function()
{
  navigator.id.watch(
  {
    loggedInUser: undefined,
    onlogin: function(assertion)
    {
      // if the user is signed in during document ready
      // this function will be called in Firefox 20,
      // Chrome 26 and even Internet Explorer 9
      // on Windows 7
    },
    onlogout: function()
    {
      // if the user is signed out during document ready
      // this function will be called in Firefox 20 and
      // Chrome 26 but _NOT_ in Internet Explorer 9
      // on Windows 7
    }
  });
});
```

Please note that I am already avoiding Internet Explorer's compatibility mode with the following code:

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
```

**Comment by jaredhirsch** [comment]
Hmm.

Tested this here: http://jsfiddle.net/XgVD5/

I'm not able to reproduce the bug. I get onlogout fired every time across IE9 and FF and Chrome.

@TeeTeeHaa In your copy of IE9 that's not firing onlogout, would you please run that fiddle and tell me what callback is fired.

However, I do see some weirdness. If I have a persona session in a given browser, this test page fires onmatch instead of onlogout, which doesn't seem to match the intent of the spec? @callahad, should we fire onmatch if the user has a persona session in that browser but not for the current RP, and loggedInUser is set to undefined?

**Comment by callahad** [comment]
@6a68 Sounds like a bug with onmatch :( onmatch should only fire when loggedInUser === my preferred email, and I have a preferred state of logged in on that site.

**Comment by TeeTeeHaa** [comment]
@6a68 Thank you for your help. I extended your Fiddle a little bit (http://jsfiddle.net/XgVD5/4/) to see whether it is actually running at all. Here are the outputs in the different browsers:

Internet Explorer 9:
1. watching...

Internet Explorer 9 of a colleague:
1. watching...

Internet Explorer 10 of a colleague:
1. watching...

Chrome 26:
1. watching...
2. onlogout fired

Firefox 21
1. watching...
2. onmatch fired

Firefox 21 of a colleague who probably has never used Persona before:
1. watching...
2. onlogout fired

**Comment by shane-tomlinson** [comment]
Could this have anything to do with: https://github.com/mozilla/browserid/pull/3194 and #3170?

**Comment by jaredhirsch** [comment]
@TeeTeeHaa sorry this went cold for a bit. Can you clarify if you did or didn't have an active persona session when you ran your tests? Additionally, can you try running that fiddle with a persona login and with a totally clean browser session? I'm still unable to reproduce your results.

**Comment by TeeTeeHaa** [comment]
All the "of a colleague" tests happened on PCs that had never used Persona before. All other tests, to be honest, I don't know anymore, sorry. Regarding further tests I'll see what I can do but I'm out of town for some days now.

**Comment by TeeTeeHaa** [comment]
Vista 64, IE9, all settings cleared, start browser, opened fiddle (http://jsfiddle.net/XgVD5/4/):

```
1.watching...
```

Clicked reload:

```
1.watching...
```

Opened another tab, opened https://login.persona.org/, clicked "Sign In", entered email address, entered password, did not have the dialog "remember always because this is your PC or remember for session only because this is public PC" (seriously, it is absolutely totally unclear for the user in which cases this dialog appears and for how long the choice is remembered), landed in account management, switched back to first tab, clicked reload:

```
1.watching...
2.onmatch fired
```

Clicked reload:

```
1.watching...
2.onmatch fired
```

Switched back to second tab, clicked "Sign Out", switched back to first tab, clicked reload:

```
1.watching...
2.onlogout fired
```

Clicked reload:

```
1.watching...
2.onlogout fired
```

Closed the second tab, clicked reload (in first tab):

```
1.watching...
2.onlogout fired
```

Closed IE 9 (Internet Options are set that everything is cleared when IE is closed), opened IE9 again, opened fiddle:

```
1.watching...
```

Set Internet Options NOT to clear everything when IE is closed, closed IE9, openen IE, opened fiddle:

```
1.watching...
```

Signed into Persona in second tab like above, switched back to first tab, clicked reload:

```
1.watching...
2.onmatch fired
```

Closed IE9, opened IE9, opened fiddle:

```
1.watching...    
2.onmatch fired   
```

Personal opinions:
- Wherever there is no "2." an "onlogout" should be fired.
- Wherever there is an "onmatch" fired there should even be an "onlogin" fired.

@6a68 Do you need this for Firefox and Chrome, too?

**Comment by jaredhirsch** [comment]
@TeeTeeHaa OK, great. Thanks for the detailed report! I'm able to reproduce this as well, after setting IE9 to delete everything at end of each browser session.

TL;DR This all goes back to cross-browser incompatibilities in third-party access to localStorage/cookies and the settings you use in your browser.

What's happening is, on initial first-run page load, login.persona.org is unable to write to localStorage, because you haven't visited the site as a first-party before, and also because we don't have a P3P header on the persona.org third-party cookies. After you visit the site (or open the dialog), persona.org is cleared to store stuff, the idea being that if you visit a site directly, you probably want whatever features they provide.

This was a known issue (#2340) with an unmerged fix (#3132) which we didn't pull in because I wasn't able to repro this problem consistently after some related third-party cookie work was merged in (#3116). After setting Win 7 to delete history on exit, though, it recurs for every new IE9 browser session, which sucks. cc @shane-tomlinson 

I'll spin up an ephemeral instance with the P3P fix and see if that resolves this bug.

Note that the same behavior (onlogout never fired) occurs if the user is browsing in Safari after doing a browser reset or in Private Browsing mode.

**Comment by jaredhirsch** [comment]
@TeeTeeHaa ok, I've updated the fiddle to use a test instance of persona that sends P3P headers to IE browsers. I bet it'll behave as expected in your IE9 sessions now, please give it a shot: http://jsfiddle.net/XgVD5/5/

**Comment by jaredhirsch** [comment]
@TeeTeeHaa finally, note that I've reopened pull request #3132 that would fix this, we'll see what @lloyd and @shane-tomlinson have to say about it in context of this bug. Thanks again for your help reporting/reproducing :beers:

**Comment by TeeTeeHaa** [comment]
IE9 set to clear everything on exit, opened IE9, opened new fiddle:

```
1.watching...
2.onlogout fired
```

Opened https://p3p-again.personatest.org in second tab, created account, clicked email link, landed in account management, clicked reload in first tab:

```
1.watching...
2.onmatch fired
```

Signed out in second tab, clicked reload in first tab:

```
1.watching...
2.onlogout fired
```

Closed IE9, opened IE9, opened new fiddle:

```
1.watching...
2.onlogout fired
```

Set IE9 not to clear everything on exit, closed IE9, opened IE9, opened new fiddle:

```
1.watching...
2.onlogout fired
```

Opened https://p3p-again.personatest.org in second tab, signed in there, clicked reload in first tab:

```
1.watching...
2.onmatch fired
```

Closed IE9, opened IE9, opened new fiddle:

```
1.watching...
2.onmatch fired
```

Good news: "onlogout" and "onmatch" seem to fire as intended now.

Bad news: "onlogin" seems not to fire in addition to "onmatch" (but I am not sure whether you actually tried to change something with your fix there).

@6a68: Thanks for your work, but could you please clarify the relation between "onmatch" and "onlogin" for me? I don't understand the documentation at https://developer.mozilla.org/en-US/docs/Web/API/navigator.id.watch which even mentions "onready".

**Comment by jaredhirsch** [comment]
@TeeTeeHaa There are two bugs here. The first bug is that none of {onlogin,onlogout,onmatch} was firing at all, because of third-party cookie policy--that seems fixed by adding P3P headers.

The second bug is that the intended meaning of onmatch (see #2648 and also [the related specs change](https://github.com/mozilla/id-specs/blob/0ad12419034273e888992305f37effbd2b4e4044/browserid/index.md#key-ideas)) doesn't match what is really happening.

'onmatch' should fire if:
- 'foo@bar.com' is already logged into an RP site, and loggedInUser == 'foo@bar.com'
- user is not logged into a given RP site, and loggedInUser == null

'onmatch' should **never** fire if loggedInUser is undefined. So, that's a separate bug I've opened as #3503.

**Comment by shane-tomlinson** [comment]
@6a68, none of onlogin, onlogout, nor onmatch should be fired if third party cookies are disabled, onready should.

To get to the point of your PR for P3P headers, they effectively enable 3rd party cookies on IE, but do _not_ fix the erroneous onmatch behavior, correct?

**Comment by jaredhirsch** [comment]
@shane-tomlinson that's correct, adding P3P headers brings IE in line with the behavior of other browsers.

**Comment by shane-tomlinson** [comment]
@6a68 - Thanks for exploring the problem and pushing on the p3p fix. p3p change merged. What do we need to do to fix the rest of this and #3503?

Can this particular issue be closed?

**Comment by TeeTeeHaa** [comment]
From my point of view this issue can be closed because it is the sum of the other two other issues.

Thanks to all for your efforts - it really encourages to submit further issues in case of problems.

How do I as a user notice that the proposed changes are live on `https://login.persona.org/include.js`?

**Comment by jaredhirsch** [comment]
w00t!

@TeeTeeHaa really appreciate your help diagnosing this in detail :beers: As far as when this code goes live, we do twice-monthly releases; this patch should go out on 2 July. More details: http://personatra.in/#2013.06.19

@shane-tomlinson You got it! The onmatch funk is covered by #3503, this one is good to close.
